From shemminger@osdl.org  Wed Jun  8 09:30:09 2005
Date: Wed, 8 Jun 2005 09:30:09 -0700
From: Stephen Hemminger <shemminger@osdl.org>
To: Chris Wright <chrisw@osdl.org>
Subject: [PATCH] [BRIDGE]: prevent bad forwarding table updates

Avoid poisoning of the bridge forwarding table by frames that have been
dropped by filtering. This prevents spoofed source addresses on hostile
side of bridge from causing packet leakage, a small but possible security
risk.

Signed-off-by: Stephen Hemminger <shemminger@osdl.org>
Signed-off-by: Chris Wright <chrisw@osdl.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 net/bridge/br_input.c    |    6 ++++--
 net/bridge/br_stp_bpdu.c |    3 +++
 2 files changed, 7 insertions(+), 2 deletions(-)

--- linux-2.6.11.10.orig/net/bridge/br_input.c	2005-05-16 10:51:56.000000000 -0700
+++ linux-2.6.11.10/net/bridge/br_input.c	2005-06-08 22:40:37.000000000 -0700
@@ -54,6 +54,9 @@
 	struct net_bridge_fdb_entry *dst;
 	int passedup = 0;
 
+	/* insert into forwarding database after filtering to avoid spoofing */
+	br_fdb_insert(p->br, p, eth_hdr(skb)->h_source, 0);
+
 	if (br->dev->flags & IFF_PROMISC) {
 		struct sk_buff *skb2;
 
@@ -108,8 +111,7 @@
 	if (eth_hdr(skb)->h_source[0] & 1)
 		goto err;
 
-	if (p->state == BR_STATE_LEARNING ||
-	    p->state == BR_STATE_FORWARDING)
+	if (p->state == BR_STATE_LEARNING)
 		br_fdb_insert(p->br, p, eth_hdr(skb)->h_source, 0);
 
 	if (p->br->stp_enabled &&
--- linux-2.6.11.10.orig/net/bridge/br_stp_bpdu.c	2005-05-16 10:51:56.000000000 -0700
+++ linux-2.6.11.10/net/bridge/br_stp_bpdu.c	2005-06-08 22:40:37.000000000 -0700
@@ -140,6 +140,9 @@
 	struct net_bridge *br = p->br;
 	unsigned char *buf;
 
+	/* insert into forwarding database after filtering to avoid spoofing */
+	br_fdb_insert(p->br, p, eth_hdr(skb)->h_source, 0);
+
 	/* need at least the 802 and STP headers */
 	if (!pskb_may_pull(skb, sizeof(header)+1) ||
 	    memcmp(skb->data, header, sizeof(header)))
